### PR TITLE
Updates Twilio Adapter to return the message sid

### PR DESCRIPTION
**Pull request by ahousand**

### CODE REVIEW

File: fixture/vcr_cassettes/twilio_sms_failure.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [
   {
     "request": {
-      "body": "Body=Hello+there&From=%2B15551234567&To=%2B15557654321",
+      "body": "Body=Hello+there&From=%2B15551234567&MessageId=&Provider=&To=%2B15557654321",
```
Comment (swelham) [inline fixture/vcr_cassettes/twilio_sms_failure.json:4]: I don't believe these two params (`MessageId` & `Provider`) are valid params to send to Twilio. I think we should drop them when constructing the params here https://github.com/seated/apus/blob/main/lib/apus/adapters/twilio_adapter.ex#L50?

File: fixture/vcr_cassettes/twilio_sms_from_success.json [hunk 2]
```diff
@@ -1,9 +1,9 @@
 [
   {
     "request": {
-      "body": "Body=Hello+there&From=%2B15551234567&To=%2B15557654321",
+      "body": "Body=Hello+there&From=%2B15551234567&MessageId=&Provider=&To=%2B15557654321",
       "headers": {
-        "Authorization": "Basic ZmFrZS1hY2NvdW50LXNpZDpmYWtlLWF1dGgtdG9rZW4=",
+        "Authorization": "Basic QUM0MGYxM2I5NTZiNjhkOGQ5NWI2NjdkMjM2YjhlOGVmZDo4Zjg0YzhhNGJhZTA5ZjIyNDE2NjQxYzIyOWZhZWVhOQ==",
```
Comment (swelham) [inline fixture/vcr_cassettes/twilio_sms_from_success.json:6]: Please revert this change.

File: fixture/vcr_cassettes/twilio_sms_from_success.json [hunk 3]
```diff
@@ -1,9 +1,9 @@
 [
   {
     "request": {
-      "body": "Body=Hello+there&From=%2B15551234567&To=%2B15557654321",
+      "body": "Body=Hello+there&From=%2B15551234567&MessageId=&Provider=&To=%2B15557654321",
```
Comment (swelham) [inline fixture/vcr_cassettes/twilio_sms_from_success.json:4]: Same comment as above about removing these new params.

File: fixture/vcr_cassettes/twilio_sms_mssid_success.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 [
   {
     "request": {
-      "body": "Body=Hello+there&From=&MessagingServiceSid=fake-mssid&To=%2B15557654321",
+      "body": "Body=Hello+there&From=&MessageId=&MessagingServiceSid=fake-mssid&Provider=&To=%2B15557654321",
```
Comment (swelham) [inline fixture/vcr_cassettes/twilio_sms_mssid_success.json:4]: Same comment as above about removing these new params.

File: lib/apus/adapters/twilio_adapter.ex [hunk 5]
```diff
@@ -20,7 +20,9 @@ defmodule Apus.TwilioAdapter do
         message = %Apus.Message{
           from: body["from"],
           to: body["to"],
-          body: body["body"]
+          body: body["body"],
+          provider: "Twilio",
```
Comment (swelham) [inline lib/apus/adapters/twilio_adapter.ex:24]: I would suggest lowercasing this to `twilio` for ease of future comparison by callers of the library.

File: fixture/vcr_cassettes/twilio_sms_from_success.json [hunk 6]
```diff
@@ -13,23 +13,25 @@
...
+      "body": "{\"body\": \"Hello there\", \"num_segments\": \"1\", \"direction\": \"outbound-api\", \"from\": \"+15551234567\", \"date_updated\": \"Tue, 06 Dec 2022 00:26:44 +0000\", \"price\": null, \"error_message\": null, \"uri\": \"/2010-04-01/Accounts/fake-account-sid/Messages/resource-id.json\", \"account_sid\": \"fake-account-sid\", \"num_media\": \"0\", \"to\": \"+15557654321\", \"date_created\": \"Tue, 06 Dec 2022 00:26:44 +0000\", \"status\": \"queued\", \"sid\": \"SM6fe74e301fe0b839d00820c31dca724e\", \"date_sent\": null, \"messaging_service_sid\": null, \"error_code\": null, \"price_unit\": \"USD\", \"api_version\": \"2010-04-01\", \"subresource_uris\": {\"media\": \"/2010-04-01/Accounts/fake-account-sid/Messages/resource-id/Media.json\"}}",
```
Comment (swelham) [inline fixture/vcr_cassettes/twilio_sms_from_success.json:16]: Can we update the `sid` value to something fake, just to make it clear it's a test value. Maybe something like `SM123`?

File: test/lib/apus/adapters/twilio_adapter_test.exs [hunk 7]
```diff
@@ -12,14 +12,22 @@ defmodule Apus.TwilioAdapterTest do
...
+          message_id: nil
+        )
 
       use_cassette "twilio_sms_from_success", match_requests_on: [:request_body] do
         {:ok, %{} = tw_message} = TwilioAdapter.deliver(message, config)
 
         assert tw_message.from == "+15551234567"
         assert tw_message.to == "+15557654321"
         assert tw_message.body == "Hello there"
+        assert !is_nil(tw_message.message_id)
```
Comment (swelham) [inline test/lib/apus/adapters/twilio_adapter_test.exs:30]: Can we assert against the `sid` value in the VCR cassette?

If not, we can simplify this slightly using a refute (ie. `refute is_nil(tw_message.message_id)`).

File: test/lib/apus/adapters/twilio_adapter_test.exs [hunk 8]
```diff
@@ -12,14 +12,22 @@ defmodule Apus.TwilioAdapterTest do
...
+          provider: nil,
+          message_id: nil
+        )
 
       use_cassette "twilio_sms_from_success", match_requests_on: [:request_body] do
         {:ok, %{} = tw_message} = TwilioAdapter.deliver(message, config)
 
         assert tw_message.from == "+15551234567"
         assert tw_message.to == "+15557654321"
         assert tw_message.body == "Hello there"
```
Comment (swelham) [inline test/lib/apus/adapters/twilio_adapter_test.exs:29]: Should we also assert the provider value is set?

File: test/lib/apus/adapters/twilio_adapter_test.exs [hunk 9]
```diff
@@ -12,15 +12,30 @@ defmodule Apus.TwilioAdapterTest do
...
+        assert tw_message.message_id == "SM123"
+        assert tw_message.provider == "twilio"
       end
+
+      cassette = File.read!("fixture/vcr_cassettes/twilio_sms_from_success.json")
+
+      assert cassette =~
+               "\"body\": \"Body=Hello+there&From=%2B15551234567&StatusCallback=https%3A%2F%2Fvalid_url.com&To=%2B15557654321\""
```
Comment (swelham) [inline test/lib/apus/adapters/twilio_adapter_test.exs:38]: What's the purpose of asserting the mock file?

Comment (ahousand) [reply test/lib/apus/adapters/twilio_adapter_test.exs:38]: purely to ensure the statuscallback gets sent back

**Comment by ahousand** [comment]
@swelham for the **life of me** I cannot figure out why changing the return in `sms_sender` is causing this, but now `assert_delivered_message` is failing (screenshot)

image

Everything is pretty much in place but this as far these changes and the corresponding API changes

**Comment by ahousand** [comment]
> @swelham for the **life of me** I cannot figure out why changing the return in `sms_sender` is causing this, but now `assert_delivered_message` is failing (screenshot)
> 
> image
> 
> Everything is pretty much in place but this as far these changes and the corresponding API changes



Update I think I've figured this out but it requires an additional case in the sms_sender

File: test/lib/apus/adapters/twilio_adapter_test.exs (hunk 9 above)
Comment (swelham) [reply test/lib/apus/adapters/twilio_adapter_test.exs:38]: I'm not sure I fully understand. The status callback will get asserted in the request by the `match_requests_on: [:request_body]` as part of the `use_cassette` macro above. If that is missing in either the cassette file or in the adapter, this test should be failing.

I don't think we need to have assertions for the static mock files, as changing the data in these files will cause tests to fail if the removed/changed data was important.

**Comment by swelham** [comment]
@ahousand Let's sync up on this when you're next online. It seems like this PR is almost ready to merge, so let's get that done and then fix the error in our app afterwards.

**Review by swelham — APPROVED** [review]
